packages/app-check/src/internal-api.ts

```diff
@@ -245,12 +255,13 @@ function createTokenRefresher(app: FirebaseApp): Refresher {
       const state = getState(app);
 
       if (state.token) {
-        return Math.max(
-          0,
-          state.token.expireTimeMillis -
-            Date.now() -
-            TOKEN_REFRESH_TIME.OFFSET_DURATION
-        );
+        // issuedAtTime + (50% * total TTL) + 5 minutes
+        const nextRefreshTimeMillis =
+          state.token.issuedAtTimeMillis +
+          (state.token.expireTimeMillis - state.token.issuedAtTimeMillis) *
+            0.5 +
+          5 * 60 * 1000;
+        return Math.max(0, nextRefreshTimeMillis - Date.now());
```

Comment:
This could put the next refresh time after the token has become expired if `state.token.expireTimeMillis - state.token.issuedAtTimeMillis < 5min`, which is wrong.

We should use whatever is smaller between `issuedAtTime + (50% * total TTL) + 5 minutes` and `expireTime - 5mins`

Reply:
Done.